yelp_scrape.py

```diff
@@ -48,6 +52,6 @@ def local_cache_check(url, file_name):
     return yelp_soup
 
 
-yelp_soup = local_cache_check(*url_generator(*get_search_request()))
+yelp_soup = local_cache_check(*url_generator())
 
```

Comment:
Okay, so the thing you're missing for the CLI stuff is the actual invocation of the CLI function.

A python file is literally just a file that executes statements one after another. Not all languages function like this, but Python does.

So you can write a file that's like
```python
print(1+1)

def foo():
  return "foo"
  
sys.exit(1)
```

when you run the file, it will execute each line in turn, it'll print "2", then define a function (that never gets called), then exit with an error (that's what sys.exit(1) does)

that's also why you can get errors like

```python
foo(3)

def foo(n: int):
  print(f"fuck the number {n}")
```

which will say "foo is not defined", even though it exists later in the file, it's because Python executes the file in-order, and it tries to execute `foo(3)` before `foo` has been defined

so with that in mind, we have to know how your CLI gets called

typically for Python programs, you write the following at the bottom of your main file

```python
if __name___ == "__main__":
  do_something()
```

what this does is check if the file is being _executed_ as a program, rather than _imported_ as a library

if the file _is_ being executed, it runs the code underneath the if
otherwise, like if you're just importing the file to get access to its function definitions and things, it won't run that code

if you don't do `if __name__ == "__main__"` you can get some *wild* bugs where like, you import a file and it executes all kinds of weird stuff during the `import` process, people don't generally expect `import x` to execute any code

if you check my CLI code, you can see I have a couple parts yours doesn't, in particular,

```python
@click.group()
def cli():
    pass
    
...
cli.add_command(convert)
...
cli.add_command(splice)
...
cli.add_command(cube)
...
cli.add_command(printing)
...

if __name__ == "__main__":
    cli()
```

The first part defines an empty function called `cli()` which just gives me something to attach the commands to (an `@click.group()`)

then I add my various subcommands to that empty function so they're accessible via the CLI

then, I check if the file is being executed, and if it is, run the `cli()` function

it looks like it doesn't do anything because `cli()` is empty, but what the `@x` style calls do is modify the function they're attached to, they're called "decorators" and they're kind of Python magic
magic is generally bad in programming, you want to understand exactly how things work so you can put stuff together, but Python has a lot of really convenient magic bullshit that people like, so these are the patterns we're stuck with, and hey, some of the magic is really useful, like Click

so yeah, you generally want your script to have one clearly marked entry point, preferably under `if __name__ == "__main__"`

if you spread statements that execute things throughout your script like some of the examples above, it can be very hard to figure out how and from where that code is getting called

if there's one `if __name__ == "__main__"` section and it's the only part that executes code, you know exactly where to go to trace your program from the moment is starts execution to the moment it ends (unless of course, you have a bunch of magic, which unfortunately Click kind of does, but it's still better to know where the non-magic stuff is happening)

Reply:
Thank you for that write up. A lot of that is information I was looking for in videos and documents but I couldn't really find a good explanation of what the @click.group() decorator really meant and why it was important. 

One question about the if `if __name__ == "__main__"` piece. When I run my program through the command line or vs code I'm writing `python yelp_scrape.py` and then it runs. Does adding the name==main piece change that at all? I saw an example of someone using name==main but their file was actually named "main.py" so I wasn't sure if my syntax would be `if __name__ == "__yelp_scrape__"` or something like that.

Reply:
Ah, that would totally make sense, but no. `main.py` is a convention (which you might want to follow), but it doesn't affect the variable stored in `__name__`.

When executed as a script,` __name__` will always be `"__main__"`
When imported, `__name__` will be the name of the module apparently (I didn't actually know that, I've never checked `__name__` for anything other than `__main__`)

Here's an article about it:
https://www.pythontutorial.net/python-basics/python-__name__/

Reply:
Yes you and that article explained the concept very well. It sounds like that's a pretty standard convention for python so I'm kind of surprised it wasn't covered in any of the Codecademy courses I've done. Maybe it was and I didn't really grasp it at the time so I forgot but either way I definitely understand better now. Thanks.

Reply:
Codecademy can't cover everything, and it tends not to cover too much relating to the "periphery" of a program
things like config files, special variables, platform-specific functionality, etc

generally when you work as a junior engineer, you don't have to deal with any of that stuff, there's a project that is already set up and you're making changes to it

I even know some senior engineers that couldn't set up a full project from scratch with like, a CLI entry point, automated testing, automated linting (code style checking), automated deployment, etc, because they've always worked somewhere where all that stuff is done for them

so Codecademy doesn't prioritize it, because at most companies it's probably already done for you

Reply:
Google for example has exactly one repository (for the most part)
there are probably less than 30 engineers at Google who have had to start a project from scratch and have to know these kinds of things

everyone else just works on the massive Google monorepository and makes changes using normal language syntax, and doesn't worry exactly how that stuff gets tested or how it gets deployed or how the program gets run or whatever